include/fmt/ranges.h

```diff
@@ -414,20 +417,86 @@ struct range_formatter<
     if (it != end && *it == 'n') {
       set_brackets({}, {});
       ++it;
+    } else {
+      bool check_for_s = false;
+      if (it != end && *it == '?') {
+        ++it;
+        detail::maybe_set_debug_format(underlying_, true);
+        set_brackets({}, {});
+        check_for_s = true;
+        is_debug = true;
+      }
+      if (it != end && *it == 's') {
+        if (!std::is_same<T, Char>::value) {
+          report_error("invalid format specifier");
+        }
+        if (!is_debug) {
+          set_brackets(detail::string_literal<Char, '"'>{},
+                       detail::string_literal<Char, '"'>{});
+        }
+        check_for_s = false;
+        is_string_format = true;
+        ++it;
+      }
+      if (check_for_s) {
+        report_error("invalid format specifier");
+      }
     }
 
     if (it != end && *it != '}') {
-      if (*it != ':') report_error("invalid format specifier");
+      if (is_string_format || *it != ':')
+        report_error("invalid format specifier");
       ++it;
     } else {
-      detail::maybe_set_debug_format(underlying_, true);
+      if (!is_string_format) detail::maybe_set_debug_format(underlying_, true);
     }
 
     ctx.advance_to(it);
     return underlying_.parse(ctx);
   }
 
-  template <typename R, typename FormatContext>
+  template <typename R, typename FormatContext, typename U = T,
+            enable_if_t<std::is_same<U, Char>::value, bool> = true>
+  auto format(R&& range, FormatContext& ctx) const -> decltype(ctx.out()) {
```

Comment:
To avoid duplication let's merge the two `format` overloads and either use if constexpr or move only the string-specific logic into a separate function.

Reply:
I tried `if constexpr` but its introduced in C++17, so tests would break. I pushed the changes to make the string logic a separate function, though I'm unsure if there's a way to cleanly handle the non-Char case, so any advice here would be appreciated.